src/github.com/matrix-org/dendrite/roomserver/input/consumer.go

```diff
@@ -0,0 +1,91 @@
+// Package input contains the code that writes
+package input
+
+import (
+	"github.com/matrix-org/dendrite/roomserver/types"
+	sarama "gopkg.in/Shopify/sarama.v1"
+)
+
+// A ConsumerDatabase has the storage APIs needed by the consumer.
+type ConsumerDatabase interface {
+	// PartitionOffsets returns the offsets the consumer has reached for each partition.
+	PartitionOffsets(topic string) ([]types.PartitionOffset, error)
+	// SetPartitionOffset records where the consumer has reached for a partition.
+	SetPartitionOffset(topic string, partition int32, offset int64) error
+}
+
+// An ErrorHandler handles the errors encountered by the consumer.
+type ErrorHandler interface {
+	OnError(err error)
+}
+
+// A Consumer consumes a kafkaesque stream of room events.
+type Consumer struct {
+	// A kafkaesque stream consumer.
+	Consumer sarama.Consumer
+	// The database used to store the room events.
+	DB ConsumerDatabase
+	// The kafkaesque topic to consume room events from.
+	RoomEventTopic string
+	// The ErrorHandler for this consumer.
+	// If left as nil then the consumer will panic with that error.
+	ErrorHandler ErrorHandler
+}
+
+// Start starts the consumer consuming.
```

Comment:
Unhelpful. What is it actually trying to do? It loads up the stored partition offsets and f.e partition it starts consuming messages in separate goroutines.